Document how to change GC logging behaviour

pugnascotia opened the pull request:
Closes #43990. Describe how to change the default GC settings without changing the default `jvm.options`, and give an example of logging to `stderr` when using Docker.

pugnascotia committed:
Document how to change GC logs in Docker

Closes #43990.

elasticmachine commented:
Pinging @elastic/es-docs (>docs)

pgomulka reviewed (approved):
LGTM

pgomulka commented:
@pugnascotia just to make sure. if someone doesn't disable gc first and just change the stream to stderr
`-Xlog:gc=debug:stderr` it wouldn't work?

pugnascotia commented:
You have to disable logging first, otherwise you get GC logs both to stderr and to file.

jrodewig commented on docs/reference/setup/important-settings/gc-logging.asciidoc:
```diff
@@ -1,7 +1,25 @@
-Elasticsearch logs. The default configuration rotates the logs every 64 MB and
-can consume up to 2 GB of disk space.
+By default, {es} enables GC logs. These are configured in
+<<jvm-options,`jvm.options`>> and default to the same default location as
```
This sounds a big redundant:
> default to the same default location

Maybe:
> **output** to the same default location

jrodewig commented on docs/reference/setup/important-settings/gc-logging.asciidoc:
```diff
@@ -1,7 +1,25 @@
-can consume up to 2 GB of disk space.
+By default, {es} enables GC logs. These are configured in
+<<jvm-options,`jvm.options`>> and default to the same default location as
+the Elasticsearch logs. The default configuration rotates the logs every 64
```
```suggestion
the {es} logs. The default configuration rotates the logs every 64
```

jrodewig commented on docs/reference/setup/important-settings/gc-logging.asciidoc:
```diff
@@ -1,7 +1,25 @@
+disable GC logging, in order to cancel out the default {es} configuration,
+and then to re-enable logging as you require. See
+https://openjdk.java.net/jeps/158[JEP 158: Unified JVM Logging] for details
+of how to configure logging.
```
I think a slight reorder and formatting would help here:

```
We do not recommend changing the default `jvm.options` file to change the GC
logging configuration. Instead, we recommend you:

. Configure the JVM to disable GC logging using the default configuration.
. Reconfigure JVM logging as described in
  https://openjdk.java.net/jeps/158[JEP 158: Unified JVM Logging].
. Re-enable GC logging.
```

It may also be helpful to briefly mention or link to:

1.  Why we discourage editing the `jvm.options` file.
2. How to disable and re-enable GC logging.

jrodewig commented on docs/reference/setup/important-settings/gc-logging.asciidoc:
```diff
@@ -1,7 +1,25 @@
+[source,sh]
+--------------------------------------------
+docker run -e ES_JAVA_OPTS="-Xlog:disable -Xlog:gc=debug:stderr" # etc
+--------------------------------------------
```
Is there a reason why this is an example vs. regular guidance?

If it could just be regular guidance, I would change as follows:

```
If you use <<docker,Docker>> to run {es}, you can send GC logs to standard error
(`stderr`) and let your container orchestrator handle the output. If using the
`ES_JAVA_OPTS` environment variable, specify:

...
```

jrodewig reviewed (approved):
LGTM overall. I left a few minor suggestions but nothing blocking. Feel free to take or ignore those as you see fit. Thanks!

pugnascotia committed:
Rework to add file example and include default config

The previous docs didn't take care to insert the default logging
configuration after disabling everything. Also, there was no example for
using a file in `jvm.options.d/`.

pugnascotia commented:
@jrodewig @pgomulka I've reworked this a fair bit, can I have another look please?

jrodewig commented on docs/reference/setup/important-settings/gc-logging.asciidoc:
```diff
@@ -1,7 +1,50 @@
+https://openjdk.java.net/jeps/158[JEP 158: Unified JVM Logging]. Unless you
+change the default `jvm.options` file directly, you may find that the {es}
+default configuration is applied in addition to your own settings. In this
+case, you must first disable logging by supplying the `-Xlog:disable`
```
```suggestion
change the default `jvm.options` file directly, the {es} default configuration
is applied in addition to your own settings. To disable the default
configuration, first disable logging by supplying the `-Xlog:disable`
```

jrodewig commented on docs/reference/setup/important-settings/gc-logging.asciidoc:
```diff
@@ -1,7 +1,50 @@
+change the default `jvm.options` file directly, you may find that the {es}
+default configuration is applied in addition to your own settings. In this
+case, you must first disable logging by supplying the `-Xlog:disable`
+option, then supply your own command line options. Note that this disables
```
```suggestion
option, then supply your own command line options. This disables
```

jrodewig commented on docs/reference/setup/important-settings/gc-logging.asciidoc:
```diff
@@ -1,7 +1,50 @@
+__all__ JVM logging, so be sure to review the available options and enable
+everything that you require.
+
+See also
```
```suggestion
See 
```

jrodewig commented on docs/reference/setup/important-settings/gc-logging.asciidoc:
```diff
@@ -1,7 +1,50 @@
+See also
+https://docs.oracle.com/en/java/javase/13/docs/specs/man/java.html#enable-logging-with-the-jvm-unified-logging-framework[Enable
+Logging with the JVM Unified Logging Framework], which includes more
+options not contained in the original JEP.
```
> , which includes more options not contained in the original JEP.

I'd inverse the order here:

> To see more options not contained in the original JEP, see ...

jrodewig commented on docs/reference/setup/important-settings/gc-logging.asciidoc:
```diff
@@ -1,7 +1,50 @@
+--------------------------------------------
+
+* Configure an {es} <<docker,Docker container>> to send GC debug logs to
+  standard error (`stderr`), in order to let the container orchestrator
```
```suggestion
  standard error (`stderr`). This lets the container orchestrator
```

pgomulka reviewed (approved):
LGTM, thank you!

jrodewig reviewed (approved):
Left some optional nits. Otherwise LGTM.

pugnascotia committed:
Address review feedback

pugnascotia committed:
One last tweak

pugnascotia commented:
Backports:

b1be7dcd2d5 `7.x`
aa751e09be0 `7.6`
